File: client/Interop/Yjs.fs
/// Thin, deliberately un-fancy bindings onto the vendored `yjs.mjs`.
///
/// This module is the ONLY place in the whole app that touches Yjs's actual JS API.
/// Everything above it (Doc.fs and up) only ever sees F# types. Keeping the surface
/// this small is on purpose: it's meant to be the file you point at during the demo
/// and say "this is the entire integration surface with the CRDT library".
///
/// Note on byte[]: Fable compiles .NET `byte[]` to a JS `Uint8Array` at runtime, which
/// is exactly what Yjs's update API expects/produces - so no manual conversion is needed
/// anywhere in this file.
module Client.Interop.Yjs

open Fable.Core
open Fable.Core.JsInterop

// Fable resolves relative import paths against this *source* file's own location
// (client/Interop/Yjs.fs) and re-relativizes them for wherever it writes the
// compiled JS - NOT against the output path itself. yjs.mjs lives in
// server/wwwroot/lib/yjs.mjs, hence "../../server/wwwroot/lib/yjs.mjs" from here.
[<ImportAll("../../server/wwwroot/lib/yjs.mjs")>]
let private Y: obj = jsNative

type YDoc =
    interface
    end

type YMap =
    interface
    end

type YText =
    interface
    end

type YArray =
    interface
    end

// -- Document ---------------------------------------------------------------

let createDoc () : YDoc = createNew Y?Doc () :?> YDoc

let getMap (name: string) (doc: YDoc) : YMap = (doc :> obj)?getMap (name)
let getArray (name: string) (doc: YDoc) : YArray = (doc :> obj)?getArray (name)

/// Marker used as the Yjs transaction "origin" for updates applied because they
/// arrived from the server, i.e. NOT a local edit. Filtering on this is what stops
/// us from immediately re-broadcasting other people's changes back to the server.
let remoteOrigin: obj = box "remote-sync"

let transact (doc: YDoc) (f: unit -> unit) : unit =
    (doc :> obj)?transact (System.Func<unit>(f)) |> ignore

/// Registers the outgoing-sync handler: fires once per local change (i.e. change
/// whose transaction origin is NOT `remoteOrigin`) with the raw update bytes Yjs
/// wants broadcast to every other client.
let onLocalUpdate (doc: YDoc) (handler: byte[] -> unit) : unit =
    (doc :> obj)?on (
        "update",
        (fun (update: byte[]) (origin: obj) ->
            if origin <> remoteOrigin then
                handler update)
    )
    |> ignore

let applyRemoteUpdate (doc: YDoc) (update: byte[]) : unit =
    Y?applyUpdate (doc, update, remoteOrigin) |> ignore

// -- Y.Map --------------------------------------------------------------------

let newMap () : YMap = createNew Y?Map () :?> YMap

let mapGetFloat (key: string) (map: YMap) : float = (map :> obj)?get (key)
let mapGetString (key: string) (map: YMap) : string = (map :> obj)?get (key)
let mapGetObj (key: string) (map: YMap) : obj = (map :> obj)?get (key)
let mapSet (key: string) (value: obj) (map: YMap) : unit = (map :> obj)?set (key, value) |> ignore
let mapDelete (key: string) (map: YMap) : unit = (map :> obj)?delete (key) |> ignore
/// Defensive existence check - lets us read fields that were added to the note "shape" after
/// some notes already existed (e.g. while iterating on this prototype in a running session).
let mapHas (key: string) (map: YMap) : bool = (map :> obj)?has (key)

let mapKeys (map: YMap) : string[] =
    JS.Constructors.Array?from ((map :> obj)?keys ())

/// Fires on any change anywhere below `map` - a field on a nested note map, or a
/// keystroke inside one of its Y.Text values. This is what lets a single observer
/// keep the whole board's read-model in sync no matter what changed.
let observeDeep (callback: unit -> unit) (map: YMap) : unit =
    (map :> obj)?observeDeep (fun (_events: obj) (_tx: obj) -> callback ()) |> ignore

// -- Y.Text ---------------------------------------------------------------------

let newText (initial: string) : YText =
    let t = createNew Y?Text () :?> YText
    (t :> obj)?insert (0, initial) |> ignore
    t

let textToString (t: YText) : string = (t :> obj)?toString ()
let textLength (t: YText) : int = (t :> obj)?length
let textInsert (index: int) (content: string) (t: YText) : unit = (t :> obj)?insert (index, content) |> ignore
let textDelete (index: int) (length: int) (t: YText) : unit = (t :> obj)?delete (index, length) |> ignore

// -- Y.Array (used for note z-order / creation order) --------------------------

let arrayPush (item: obj) (arr: YArray) : unit = (arr :> obj)?push ([| item |]) |> ignore

let arrayToList (arr: YArray) : string list =
    (arr :> obj)?toArray () |> unbox<string[]> |> List.ofArray

let arrayRemoveValue (value: string) (arr: YArray) : unit =
    let items = (arr :> obj)?toArray () |> unbox<string[]>
    let idx = Array.tryFindIndex ((=) value) items

    match idx with
    | Some i -> (arr :> obj)?delete (i, 1) |> ignore
    | None -> ()

// -- Y.UndoManager --------------------------------------------------------------

type YUndoManager =
    interface
    end

/// `scope` is one or more shared types whose changes - including nested content below them,
/// e.g. fields inside a Y.Map living inside this one - should be tracked. By default only
/// local changes are tracked (transactions with no explicit origin), which is exactly the
/// complement of `remoteOrigin` above: undo() can therefore never touch another user's change.
let newUndoManager (scope: obj[]) : YUndoManager =
    createNew Y?UndoManager scope :?> YUndoManager

let undo (um: YUndoManager) : unit = (um :> obj)?undo () |> ignore
let redo (um: YUndoManager) : unit = (um :> obj)?redo () |> ignore
let canUndo (um: YUndoManager) : bool = (um :> obj)?canUndo ()
let canRedo (um: YUndoManager) : bool = (um :> obj)?canRedo ()

File: server/wwwroot/js/Interop/Yjs.js

import * as yjs from "../../lib/yjs.mjs";
import { equals } from "../fable_modules/fable-library-js.5.13.0/Util.js";
import { ofArray } from "../fable_modules/fable-library-js.5.13.0/List.js";
import { tryFindIndex } from "../fable_modules/fable-library-js.5.13.0/Array.js";

export function createDoc() {
    return new yjs.Doc();
}

export function getMap(name, doc) {
    return doc.getMap(name);
}

export function getArray(name, doc) {
    return doc.getArray(name);
}

export const remoteOrigin = "remote-sync";

export function transact(doc, f) {
    doc.transact(f);
}

/**
 * Registers the outgoing-sync handler: fires once per local change (i.e. change
 * whose transaction origin is NOT `remoteOrigin`) with the raw update bytes Yjs
 * wants broadcast to every other client.
 */
export function onLocalUpdate(doc, handler) {
    doc.on("update", ((update, origin) => {
        if (!equals(origin, remoteOrigin)) {
            handler(update);
        }
    }));
}

export function applyRemoteUpdate(doc, update) {
    yjs.applyUpdate(doc, update, remoteOrigin);
}

export function newMap() {
    return new yjs.Map();
}

export function mapGetFloat(key, map) {
    return map.get(key);
}

export function mapGetString(key, map) {
    return map.get(key);
}

export function mapGetObj(key, map) {
    return map.get(key);
}

export function mapSet(key, value, map) {
    map.set(key, value);
}

export function mapDelete(key, map) {
    map.delete(key);
}

/**
 * Defensive existence check - lets us read fields that were added to the note "shape" after
 * some notes already existed (e.g. while iterating on this prototype in a running session).
 */
export function mapHas(key, map) {
    return map.has(key);
}

export function mapKeys(map) {
    return Array.from(map.keys());
}

/**
 * Fires on any change anywhere below `map` - a field on a nested note map, or a
 * keystroke inside one of its Y.Text values. This is what lets a single observer
 * keep the whole board's read-model in sync no matter what changed.
 */
export function observeDeep(callback, map) {
    map.observeDeep((_events, _tx) => {
        callback();
    });
}

export function newText(initial) {
    const t = new yjs.Text();
    t.insert(0, initial);
    return t;
}

export function textToString(t) {
    return t.toString();
}

export function textLength(t) {
    return t.length | 0;
}

export function textInsert(index, content, t) {
    t.insert(index, content);
}

export function textDelete(index, length, t) {
    t.delete(index, length);
}

export function arrayPush(item, arr) {
    arr.push([item]);
}

export function arrayToList(arr) {
    return ofArray(arr.toArray());
}

export function arrayRemoveValue(value, arr) {
    const idx = tryFindIndex((y) => (value === y), arr.toArray());
    if (idx == null) {
    }
    else {
        const i = idx | 0;
        arr.delete(i, 1);
    }
}

/**
 * `scope` is one or more shared types whose changes - including nested content below them,
 * e.g. fields inside a Y.Map living inside this one - should be tracked. By default only
 * local changes are tracked (transactions with no explicit origin), which is exactly the
 * complement of `remoteOrigin` above: undo() can therefore never touch another user's change.
 */
export function newUndoManager(scope) {
    return new yjs.UndoManager(scope);
}

export function undo(um) {
    um.undo();
}

export function redo(um) {
    um.redo();
}

export function canUndo(um) {
    return um.canUndo();
}

export function canRedo(um) {
    return um.canRedo();
}

